drivers: dp: use atomic APIs for pin set/reset

fabiobaltieri opened the pull request:
Use LL_GPIO_SetOutputPin and LL_GPIO_ResetOutputPin for the STM32 optimized DP functions. This yelds a speedup of the bit-banged interface from about 585kHz to 640kHz on an STM32C0.

fabiobaltieri committed:
drivers: dp: use atomic APIs for pin set/reset

Use LL_GPIO_SetOutputPin and LL_GPIO_ResetOutputPin for the STM32
optimized DP functions. This yelds a speedup of the bit-banged interface
from about 585kHz to 640kHz on an STM32C0.

Suggested-by: Mathieu Choplain <mathieu.choplain@st.com>
Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

fabiobaltieri commented:
Hi @wearyzen, I'm afraid I'll need your help again on some assembly stuff... so I did https://github.com/zephyrproject-rtos/zephyr/pull/89750/files by adding a `.syntax unified` to the snippet and it worked alright, now I swapped a couple of calls from the STM32 in the hal in this PR and all of a sudden I get a build error:

```
/tmp/ccSNeDOS.s: Assembler messages:
/tmp/ccSNeDOS.s:773: Error: cannot honor width suffix -- `movs r3,ip'
/tmp/ccSNeDOS.s:853: Error: cannot honor width suffix -- `movs r3,ip'
/tmp/ccSNeDOS.s:1458: Error: cannot honor width suffix -- `movs r3,ip'
/tmp/ccSNeDOS.s:3687: Error: cannot honor width suffix -- `movs r3,ip'
/tmp/ccSNeDOS.s:4578: Error: cannot honor width suffix -- `movs r3,ip'
/tmp/ccSNeDOS.s:4656: Error: cannot honor width suffix -- `movs r3,ip'
```

Which is very very weird as it seems like the last two of those belong to the `swdp_ll_pin_get` code which I did not change.

If I remove the assembly snippet in `drivers/dp/swdp_ll_pin.h` then it builds again, and it also works if remove the unified statement in the loop and replace the `subs` with a `sub` and `cmp`. Tried to add a `.syntax divided` as well as to set the `zephyr_library_compile_options(-masm-syntax-unified)` for the whole dp library and made no difference.

`LL_GPIO_SetOutputPin` expands to simply

```c
static inline void LL_GPIO_ResetOutputPin(GPIO_TypeDef *GPIOx, uint32_t PinMask)
{
  ((GPIOx->BRR) = (PinMask));
} 
```

so nothing fancy here, but on the assembly file I can see indeed that there's a bunch of blocks like this

```asm
@ 14 "/home/fabiobaltieri/zephyrproject/zephyr/drivers/dp/swdp_ll_pin.h" 1
        .syntax unified
movs r3, ip
.start_768:
subs r3, #1
bne .start_768
```

Though I'm not sure what to make out of this. Any ideas?

jfischer-no commented on drivers/dp/swdp_ll_pin_stm32.h:
```diff
@@ -35,29 +35,15 @@ static ALWAYS_INLINE void swdp_ll_pin_output(void *const base, uint8_t pin)
-	LL_GPIO_WriteOutputPort(gpio, val);
-
-	z_stm32_hsem_unlock(CFG_HW_GPIO_SEMID);
+	LL_GPIO_SetOutputPin(gpio, BIT(pin));
```
"atomic APIs" is confusing, I do not see it using any API. Actually, I thing it should not be obfuscated but written out like for the nRF GPIO controller.

`GPIOx->BSRR = BIT(pin);`

`GPIOx->BSR = BIT(pin);` / `GPIOx->BSRR = BIT(pin << 16);`

So, we know how many CPU cycles it takes.

mathieuchopstm replied:
Raw hardware access will require a ton of `#if (SOC_SERIES_XXX) || (SOC_SERIES_YYY)` which `LL_GPIO_SetOutputPin`/`ClearOutputPin` hides at no additional cost.

fabiobaltieri replied:
Fine by me, added a commit, still wondering about the assembly error though. :-/

mathieuchopstm replied:
\+ `BIT(pin + 16)`, not `BIT(pin << 16)`

fabiobaltieri replied:
Uh-oh


```c
stm32cube/stm32g4xx/drivers/include/stm32g4xx_ll_gpio.h:__STATIC_INLINE void LL_GPIO_ResetOutputPin(GPIO_TypeDef *GPIOx, uint32_t PinMask)
stm32cube/stm32g4xx/drivers/include/stm32g4xx_ll_gpio.h-{
stm32cube/stm32g4xx/drivers/include/stm32g4xx_ll_gpio.h-  WRITE_REG(GPIOx->BRR, PinMask);
stm32cube/stm32g4xx/drivers/include/stm32g4xx_ll_gpio.h-}
```

```
stm32cube/stm32h7xx/drivers/include/stm32h7xx_ll_gpio.h:__STATIC_INLINE void LL_GPIO_ResetOutputPin(GPIO_TypeDef *GPIOx, uint32_t PinMask)
stm32cube/stm32h7xx/drivers/include/stm32h7xx_ll_gpio.h-{
stm32cube/stm32h7xx/drivers/include/stm32h7xx_ll_gpio.h-  WRITE_REG(GPIOx->BSRR, PinMask << 16U);
stm32cube/stm32h7xx/drivers/include/stm32h7xx_ll_gpio.h-}
```

nah, not going to work

fabiobaltieri replied:
Yeah looks like some stm32 use separate register and some the high/low nibble. I don't think we want to bring that in here.

fabiobaltieri replied:
> + `BIT(pin + 16)`, not `BIT(pin << 16)`

Looks like it depends on the specific part, I think we are better off with the obfuscated call, don't want to start enumerating all possible variants here.

jfischer-no replied:
We either know what is going on, or we use the GPIO API. I am confident that this could be implemented at a very low level in a readable and maintainable form. It was my mistake to approve the initial PR, will not do it again but I will not block this change.

mathieuchopstm commented:
> ```
> /tmp/ccSNeDOS.s: Assembler messages:
> /tmp/ccSNeDOS.s:773: Error: cannot honor width suffix -- `movs r3,ip'
> /tmp/ccSNeDOS.s:853: Error: cannot honor width suffix -- `movs r3,ip'
> /tmp/ccSNeDOS.s:1458: Error: cannot honor width suffix -- `movs r3,ip'
> /tmp/ccSNeDOS.s:3687: Error: cannot honor width suffix -- `movs r3,ip'
> /tmp/ccSNeDOS.s:4578: Error: cannot honor width suffix -- `movs r3,ip'
> /tmp/ccSNeDOS.s:4656: Error: cannot honor width suffix -- `movs r3,ip'
> ```

I didn't check on which targets it fails to build, but my gut feeling is that it's ARMv6-M and probably caused by `movs <Rt>, <HiReg>` not existing on these cores? (maybe `mov r3, ip`, without the S, would work instead?)

fabiobaltieri commented:
> I didn't check on which targets it fails to build, but my gut feeling is that it's ARMv6-M and probably caused by `movs <Rt>, <HiReg>` not existing on these cores? (maybe `mov r3, ip`, without the S, would work instead?)

Yeah that works on H7, fails on C0

`west build -p -b nucleo_h745zi_q/stm32h745xx/m7 samples/subsys/dap` good
`west build -p -b nucleo_c071rb samples/subsys/dap` bad

I presume it'd work but I'd expect the compiler to do the right thing here

mathieuchopstm commented:
> > I didn't check on which targets it fails to build, but my gut feeling is that it's ARMv6-M and probably caused by `movs <Rt>, <HiReg>` not existing on these cores? (maybe `mov r3, ip`, without the S, would work instead?)
> 
> Yeah that works on H7, fails on C0
> 
> `west build -p -b nucleo_h745zi_q/stm32h745xx/m7 samples/subsys/dap` good `west build -p -b nucleo_c071rb samples/subsys/dap` bad
> 
> I presume it'd work but I'd expect the compiler to do the right thing here

AFAIU, GCC treats inline ASM blocks as opaque and relies only on operands information to fill in the gaps. If the inline ASM cannot work with the contraints specified in operands, GCC is not at fault :slightly_smiling_face: 

According to https://gcc.gnu.org/onlinedocs/gcc/Machine-Constraints.html, I think using `l` instead of `r` would fix the issue; but replacing `movs` with `mov` is more sensible as it gives more freedom to GCC... but actually, the optimal solution might be:
```c
    __asm__ volatile(
        ".syntax unified\n"
		".start_%=:\n"
		"subs %0, #1\n"
		"bne .start_%=\n"
        : "+l" (delay)
        :
        : "cc"
    );
```
(https://godbolt.org/z/K1PK4a4Kf)

This gives GCC total freedom of which register to use for the loop, how to load the value into it, etc, all while guaranteeing it works on all cores due to the `l` constraint. (Note how `delay` is declared as a RW Output rather than Input to allow usage in-place; Godbolt link demonstrates why Input is not correct).

fabiobaltieri committed:
drivers: dp: rework the ARM delay loop

Rework the ARM delay loop to avoid it causing assembler issue in the
rest of the code.

Link: https://github.com/zephyrproject-rtos/zephyr/pull/90211#issuecomment-2894157705
Suggested-by: Mathieu Choplain <mathieu.choplain@st.com>
Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

fabiobaltieri commented:
> According to https://gcc.gnu.org/onlinedocs/gcc/Machine-Constraints.html, I think using `l` instead of `r` would fix the issue; but replacing `movs` with `mov` is more sensible as it gives more freedom to GCC... but actually, the optimal solution might be:

Cool! That works, didn't even try to explain it in the commit just linked your comment. Tested, works, many thanks for the suggestion.

mathieuchopstm reviewed (approved):
LGTM if CI approves

fabiobaltieri commented:
Cool, @jfischer-no let me know if you feel strongly about that inline expansion, I could check the hal and come up with a big ifdef for the two variations but honestly I'd rather not. I don't like the hal either but it's there already, may as well keep using it.

fabiobaltieri commented:
thanks @wearyzen, all good, saw the reference that Mathieu posted, makes a lot of sense, need to put some time to familiarize with asm snippets format, I don't get to do it much myself, appreciate the support though!
